search/shortest.go

```diff
@@ -0,0 +1,291 @@
+// Copyright ©2014 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package search
+
+import (
+	"math"
+	"math/rand"
+
+	"github.com/gonum/graph"
+	"github.com/gonum/matrix/mat64"
+)
+
+// Shortest is a shortest-path tree created by the DijkstraFrom function.
+type Shortest struct {
+	// from holds the source node given to
+	// DijkstraFrom.
+	from graph.Node
+
+	// nodes hold the nodes of the analysed
+	// graph.
+	nodes []graph.Node
+	// indexOf contains a mapping between
+	// the id-dense representation of the
+	// graph and the potentially id-sparse
+	// nodes held in nodes.
+	indexOf map[int]int
+
+	// dist and next represent the shortest
+	// paths between nodes.
+	//
+	// Indices into dist and next are
+	// mapped through indexOf.
+	//
+	// dist contains the distances
+	// from the from node for each
+	// node in the graph.
+	dist []float64
+	// next contains the shortest-path
+	// tree of the graph. The index is a
+	// linear mapping of from-dense-id
+	// and to-dense-id.
+	next []int
+}
+
+func (p Shortest) set(to int, weight float64, mid int) {
+	p.dist[to] = weight
+	p.next[to] = mid
+}
+
+// From returns the starting node of the paths held by the Shortest.
+func (p Shortest) From() graph.Node { return p.from }
+
+// WeightTo returns the weight of the minimum path to v.
+func (p Shortest) WeightTo(v graph.Node) float64 {
+	to, toOK := p.indexOf[v.ID()]
+	if !toOK {
+		return math.Inf(1)
+	}
+	return p.dist[to]
+}
+
+// To returns a shortest path to v and the weight of the path.
+func (p Shortest) To(v graph.Node) (path []graph.Node, weight float64) {
+	to, toOK := p.indexOf[v.ID()]
+	if !toOK {
+		return nil, math.Inf(1)
+	}
+	from := p.indexOf[p.from.ID()]
+	path = []graph.Node{v}
+	for to != from {
+		path = append(path, p.nodes[p.next[to]])
+		to = p.next[to]
+	}
+	for i, j := 0, len(path)-1; i < j; i, j = i+1, j-1 {
+		path[i], path[j] = path[j], path[i]
+	}
+	return path, p.dist[p.indexOf[v.ID()]]
+}
+
+// ShortestPaths is a shortest-path tree created by the FloydWarshall function.
+type ShortestPaths struct {
+	// nodes hold the nodes of the analysed
+	// graph.
+	nodes []graph.Node
+	// indexOf contains a mapping between
+	// the id-dense representation of the
+	// graph and the potentially id-sparse
+	// nodes held in nodes.
+	indexOf map[int]int
+
+	// dist, next and forward represent
+	// the shortest paths between nodes.
+	//
+	// Indices into dist and next are
+	// mapped through indexOf.
+	//
+	// dist contains the pairwise
+	// distances between nodes.
+	dist *mat64.Dense
+	// next contains the shortest-path
+	// tree of the graph. The first index
+	// is a linear mapping of from-dense-id
+	// and to-dense-id; the second is the
+	// intermediate nodes by dense id.
+	// The interpretation of next is
```

Comment:
The interpretation of next depending on forward is ok, I think. But it seems to me that the second sentence is lacking in clarity. I find `The first index is a linear mapping of from-dense-id and to-dense-id; the second is the intermediate nodes by dense id` slightly cryptic. The first part talks about index and without looking at the code I could imagine what it means. In the second part, however, "the second" seems to refer not to index (because the index does not matter) but to the contents of next[linmap(from,to)]. Also, I am not sure how well "intermediate nodes" transmits the contents of next[linmap(from,to)], although it will be a standard graph terminology, I guess.